lib/jcsda-emc/spack-stack/stack/stack_env.py

```diff
@@ -187,7 +187,9 @@ def write(self):
         # Commonly used config settings
         if self.compiler:
             compiler = 'packages:all::compiler:[{}]'.format(self.compiler)
```

Comment:
If we keep lines 189 and 191, we need to remove them from all site configs. What happens for new site configs if you don't use lines 189 and 191, but just the new logic? It would be nice for the preconfigured sites to have the full list of available/possible compilers specified in `packages:all:compilers`, and then use the `require` line to switch.

Reply:
I'm not sure I follow-- Why would the compilers need to get removed from site configs? It's arguably redundant to have, say, `packages:all:compiler:[intel,gcc]` in my packages.yaml _and_ `spack:packages:all:compiler:[intel]` in my spack.yaml, but I don't think it hurts anything.

Reply:
Correct, it doesn't hurt but it's confusing. If you look at https://github.com/NOAA-EMC/spack-stack/blob/develop/configs/sites/cheyenne/packages.yaml, we use double colons for `packages:all:compiler::`, which means overwrite any other configuration entry of the same time (i.e. do not concatenate lists). In other words, what the spack stack create env script creates for `packages:all:compiler:` is ignored. I can imagine this being confusing for users.

Looking at our documentation, we explicitly tell users who configure new sites to run the spack add commands for both compiler and MPI. The compiler command wouldn't be required anymore if we kept what's in the spack extension at the moment, but the MPI command would still be required. Not ideal, but ok.

I don't have any better idea than what you already have, so we might as well try this and see how it goes.

Reply:
I like having the compiler option in this PR. I'm always having to modify the `spack add ` command for the compiler in the instructions with the actual version I have on my Mac. 

Would it make sense to add to this PR an option to specify the MPI in the `spack create` command. That way both the compiler and MPI add commands in the current instructions are not required. Seems that this might be a little less confusing.

Reply:
I like that, and was considering that except I wasn't sure how it would work. With the "one_of" logic, it's easy enough to ensure it uses only that one compiler, but if there were multiple MPI providers listed in sites/packages.yaml, I'm not sure how to restrict it to just one. If there's a way to make the logic work then I'm definitely happy to include that.

Reply:
That's actually pretty simple (but I have some other concerns):
- for external packages, it's done via the spec definition, e.g. `intel-oneapi-mpi@2021.4.0%intel@2021.5.1` for example. This way, when you specify the compiler, the MPI library is fixed (as long as you use each compiler with only one MPI library in the site config). As long as all MPI providers are in the MPI provider list, this will work. The disadvantage of adding it at the command line is that, in addition of having to know the exact name of the compiler@version, the user needs to know the MPI library, and specifying it doesn't make any sense, because there is only one choice for each compiler anyway.
- for MPI being built by spack it's a different story. Whatever MPI you choose, it's going to built it with the compiler you want (because of require / one_of).

I actually have a different idea that may make life a lot easier (I think, if it works). Let's hold off merging this PR for now until I had a chance to try, ok?

Reply:
Yeah that's fine by me, this isn't super urgent.

That said, I'm not sure I'm too worried about the need to know the exact version etc. since this argument is optional, and my anticipated use for this would be a situation like installing on, say, Acorn, where we definitely know what compiler and MPI we want for a given installation and want to be 100% explicit/exact about it.

Reply:
BTW I like the idea of having only the "one_of" which works fine as far as spack.yaml/concretization, but for some reason the command doesn't work when I remove those lines. I'll keep looking.

Reply:
I haven't looked at this for a while (sorry). I'm under the impression that we are still waiting to experiment with @climbfuji's idea which is fine. I'm just checking to make sure you are not waiting for me. Thanks!